### PR TITLE
Schema support

### PR DESCRIPTION
**Pull request by zigomir**
#9

I don't have any tests yet because I didn't think about a nice way to do it yet, but I think early CR is always good.

I don't know what's wrong with GitHub's diff though.

### CODE REVIEW

File: bin/schemats.ts [hunk 1]
```diff
@@ -18,10 +18,12 @@ let argv: any = yargs
     .alias('c', 'conn')
     .nargs('c', 1)
     .describe('c', 'database connection string')
-    .demand('t')
```
Comment (xiamx) [inline bin/schemats.ts]: I think we need to check if neither table parameter nor schema parameter is provided by the user. In this case we can print a help message letting the user know that at least one of these parameters need to be provided.

Comment (zigomir) [inline bin/schemats.ts]: I agree. But then I was thinking if no schema nor tables are provided, we could output all tables from default (public) schema. What do you think?

Comment (xiamx) [inline bin/schemats.ts]: That works too. Just need to document this behaviour in `schemats --help` output.

**Comment by zigomir** [comment]
@xiamx please take another look. I added the tests. Current algo is like such:

If tables are provided only generate those tables (ignore schema option).
If schema is provided generate all tables in schema.
If no schema or tables are provided generate tables from public schema.

File: test/test.ts [hunk 2]
```diff
@@ -3,49 +3,80 @@
...
-import {loadOSM} from './load_osm'
+import { loadSchema } from './load_schema'
 import * as fs from 'mz/fs'
 import { typescriptOfSchema, Database, extractCommand } from '../src/index'
 import * as diff from 'diff'
 
+async function compare(outputFile: string, formattedOutput: { dest: string }) {
+    await fs.writeFile(outputFile, formattedOutput.dest)
+
+    let gold = await fs.readFile('./test/example/osm.ts', {encoding: 'utf8'});
```
Comment (xiamx) [inline test/test.ts]: should this be an argument for function `compare`?

**Comment by xiamx** [comment]
Thanks for the updates @zigomir . 

I'm a little concerned about the terminologies of `schema` causing confusion in understanding the code. I used the term _schema_ in the sense of [database schema](https://en.wikipedia.org/wiki/Database_schema). But Postgres uses _schema_ as a named collection of tables. Any suggestions for unambiguous namings? 

Perhaps @vitaly-t can help us out. 


@pdelanauze @ecoutu

File: test/test.ts [hunk 3]
```diff
@@ -3,49 +3,80 @@
...
-                const t = d.added ? '+' : d.removed ? '-' : 'x';
-                console.error(`  [${i}] ${t} ${d.value}`);
-            })
-            process.exit(1)
-        } else {
-            console.log('Generated type definition identical to the standard')
-            process.exit(0)
-        }
-
+        testGeneratingTables(db)
```
Comment (xiamx) [inline test/test.ts]: need to `await` this

File: test/test.ts [hunk 4]
```diff
@@ -3,49 +3,80 @@
...
+        testGeneratingSchema(db)
```
Comment (xiamx) [inline test/test.ts]: here too

**Comment by vitaly-t** [comment]
>I'm a little concerned about the terminologies of schema causing confusion in understanding the code. I used the term schema in the sense of database schema. But Postgres uses schema as a named collection of tables. Any suggestions for unambiguous namings?

Schema = namespace of tables :wink:

**Comment by jmealo** [comment]
I'd like to take a stab at ENUM support and comments. It seems like we should merge this first though.

**Comment by xiamx** [comment]
> I can take a stab at ENUM support and/or comments. It seems like we should merge this first though. 

@jmealo yes, this should be merged in first. I think the latest commits in the PR are worth merging in. Just need to wait for @zigomir 's response regarding the latest review comments. 

That said, I don't think the internal code will change much. So if you can't wait to get started, you can fork this repo and merge in @zigomir's current work in progress.

**Comment by abenhamdine** [comment]
Hi and thx all for this great module...
Can't wait for this must-have feature !

**Comment by zigomir** [comment]
@xiamx thanks for picking this up. I didn't have any time or motive to work on this for the last days…let me know if I can help you with anything.

**Comment by xiamx** [comment]
No problem @zigomir , thanks for taking the initiative!

**Pull request merged**